Add typescript typings

Reewr opened the pull request:
This will allow typescript users of this package to get full type
information as well as autocompletion should they use such a tool.

There are no actual code changes with this commit and should have no
impact on the tests themselves.

Note: Ideally, `@types/node` should have been a dependency in this
package as the `InspectOptions` type is defined within this package.
However, as that was the only type that was used, I decided to copy it
from the type package instead. The downside of this is that if
`InspectOptions` is updated, the index.d.ts won't automatically be
updated.

Another note: At the bottom of the index.d.ts, it exposes four different
versions of the test function. However, these are only a few of the
different versions that can be used as the arguments can be sent to it
in any order.

## Test the types

A simple way to test that this works is:

```bash
# first clone the repo with the changes
git clone git@github.com:reewr/awfltst.git awfltst-reewr
cd awfltst-reewr
git checkout ts-typings

# now create a test folder for new npm project
cd ..
mkdir test-ts-typings
cd test-ts-typings
npm init

# install typescript for testing
npm install --save typescript
npm install ../awfltst-reewr # this creates a symlink to the repo
./node_modules/.bin/tsc --init

# Edit the newly generated tsconfig.json file and update `target` to `es6`
# as it uses async functions (async functions are a part of TypeScript, 
# but it requires ES6 to function due to needing Promises)
```
Add a new file:
```typescript
// index.ts
import test from 'awfltst';

test('something', async function() {
    this.eq(5, 1);
    this.eq(1, 1, 'message', 'notAnArgument'); // this line should have an error
    this.approx('string', 5, 1); // and this, because only numbers are allowed
});
```

You should now be able to run `tsc` and see the following errors due to incorrect types:

```text
index.ts:5:5 - error TS2554: Expected 2-3 arguments, but got 4.

5     this.eq(1, 1, 'message', 'notAnArgument'); // this line should have an error
      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

index.ts:6:17 - error TS2345: Argument of type '"string"' is not assignable to parameter of type 'number'.

6     this.approx('string', 5, 1);  // and this, because only numbers are allowed
                  ~~~~~~~~


Found 2 errors.
```

Reewr committed:
Add typescript typings

This will allow typescript users of this package to get full type
information as well as autocompletion should they use such a tool.

There are no actual code changes with this commit and should have no
impact on the tests themselves.

Note: Ideally, `@types/node` should have been a dependency in this
package as the `InspectOptions` type is defined within this package.
However, as that was the only type that was used, I decided to copy it
from the type package instead. The downside of this is that if
`InspectOptions` is updated, the index.d.ts won't automatically be
updated.

Another note: At the bottom of the index.d.ts, it exposes four different
versions of the test function. However, these are only a few of the
different versions that can be used as the arguments can be sent to it
in any order.

Thhethssmuz commented:
Nice :D

I don't actually mind having `@types/node` as a dependency. Would probably be preferable to manually updating this at some future point... So feel free to add that.

I did find a few issues though:

---

I assume you used the JSDocs to generate this. However, after testing a little, I have found a few cases where the documentation itself is wrong...

- `this.contains(actual, expected[, name])`
- `this.notContains(actual, expected[, name])`
- `this.in(actual, expected[, name])`
- `this.notIn(actual, expected[, name])`

which currently blocks one of the intended use cases for these functions:

```typescript
import test from 'awfltst';
test('types', async function() {
  this.contains([1, 2, 3], 2);
  this.notContains([1, 2, 3], 2);
  this.in(3, [1, 2, 3]);
  this.notIn(3, [1, 2, 3]);
});
```

```
index.ts:3:28 - error TS2345: Argument of type '2' is not assignable to parameter of type 'string | any[]'.

3   this.contains([1, 2, 3], 2);
                             ~

index.ts:4:8 - error TS2339: Property 'notContains' does not exist on type 'Test'.

4   this.notContains([1, 2, 3], 2);
         ~~~~~~~~~~~

index.ts:5:11 - error TS2345: Argument of type '3' is not assignable to parameter of type 'string | any[]'.

5   this.in(3, [1, 2, 3]);
            ~

index.ts:6:14 - error TS2345: Argument of type '3' is not assignable to parameter of type 'string | any[]'.

6   this.notIn(3, [1, 2, 3]);
               ~
```

These functions should probably all be updated to `actual: any` and `expected: any`. I'll see if I can update the documentation asap..

... also type definition is missing for `notContains`.

---

The functions `this.throws` and `this.notThrows` have two optional arguments, however, type definition is wrong when only providing the last optional argument.

- `this.throws(test[, expected][, name])`
- `this.notThrows(test[, expected][, name])`

```typescript
import test from 'awfltst';
test('types', async function() {
  this.throws(async function () {}, 'should throw');
  this.notThrows(async function () {}, 'should not throw');
});
```

```
index.ts:3:37 - error TS2345: Argument of type '"should throw"' is not assignable to parameter of type 'Function | RegExp | undefined'.

3   this.throws(async function () {}, 'should throw');
                                      ~~~~~~~~~~~~~~

index.ts:4:40 - error TS2345: Argument of type '"should not throw"' is not assignable to parameter of type 'Function | RegExp | undefined'.

4   this.notThrows(async function () {}, 'should not throw');
                                         ~~~~~~~~~~~~~~~~~~
```

---

Finally, not an issue, but a question...

My knowledge of TypeScript is rather limited, but is it possible to define type dependencies? 

For example, would it be possible to define `this.eq(actual, expected[, name])` to `actual: any` and `expected: any`, but require that `actual` and `expected` is of the same type?

It sounds to me like we can utilize TypeScript here to catch errors in tests at "compile-time" rather than having to wait until runtime and find a failed test... However, as I said, my knowledge of TypeScript is limited, and I have no idea if this would work as I imagine.

Reewr committed:
Add `@types/node` as dependency

This removes the declaration of the InspectOptions retrieved from the
type library and instead adds a dependency of it. This means that
there's no need to manually update the InspectOptions declaration should
it change in the NodeJS API.

Reewr commented:
> I don't actually mind having `@types/node` as a dependency. Would probably be preferable to manually updating this at some future point... So feel free to add that.
> 

Added as its own commit.

Regarding the functions, is the use case below the way it should always be used? `contains` and `notContains` always takes the array as first argument, whereas `in` and `notIn` takes the array as the second argument?

```typescript
contains(actual: Array<T>, expected: T, name?: string): Test;
contains(actual: string, expected: string, name?: string): Test;

in(actual: T, expected: Array<T>, name?: string): Test;
in(actual: string, expected: string, name?: string): Test;
```
> import test from 'awfltst';
> test('types', async function() {
>   this.contains([1, 2, 3], 2);
>   this.notContains([1, 2, 3], 2);
>   this.in(3, [1, 2, 3]);
>   this.notIn(3, [1, 2, 3]);
> });

> The functions `this.throws` and `this.notThrows` have two optional arguments, however, type 
definition is wrong when only providing the last optional argument.

I'll fix this by adding a separate function declaration for these.

> Finally, not an issue, but a question...
> 
> My knowledge of TypeScript is rather limited, but is it possible to define type dependencies?
> 
> For example, would it be possible to define `this.eq(actual, expected[, name])` to `actual: any` and `expected: any`, but require that `actual` and `expected` is of the same type?
> 

Yes you can, unlike similarly to C++, but they're called [Generics](https://www.typescriptlang.org/docs/handbook/generics.html)

> It sounds to me like we can utilize TypeScript here to catch errors in tests at "compile-time" rather than having to wait until runtime and find a failed test... 

Exactly! An example would be `typeof`, where the `string` argument could be an [Enum](https://www.typescriptlang.org/docs/handbook/enums.html) that matches either `bool`, `string`, `number` and so on, since the `typeof`-operator has only a very limited amount of options.:

Reewr committed:
Allow throws/notThrows with a message only

Reewr committed:
Fix definition of `contains` and `in`.

Also adds `notContains`. This uses generics to fix this issue as that
gives you typechecking on both arguments

Reewr commented:
I see that `contains` and `in` are not entirely working as I had hoped.

```typescript
contains<T>(actual: T[]|string, expected: T, name?: string): Test;

in<T>(actual: T, expected: T[]|string, name?: string): Test;
```

For instance, it correctly handles these:

```typescript
this.contains('something', '1');
this.contains([1, 2, 3, 4], 5);
this.contains([new Date(), new Date(), new Date()], new Date());
this.in(new Date(), [new Date(), new Date(), new Date()]);
```

However, the definition also allows:

```typescript
this.contains('something', 5);
this.contains('something', new Date());
```

The first could be okay, the second, not so much.

Reewr committed:
Fix issue with `contains` and `in`

The issue was that since strings was the other acceptable argument, you
could send in `5` and `"1234" and the typechecking would allow it. This
new definition only allows you to send in strings when you are either
using a string or an array of strings.

Reewr commented:
I may have fixed it. The new definition:

```typescript
type ArrayLike<T> =
  T extends string ? string | Array<string> :
  T extends (infer U)[] ? Array<U> :
  Array<T>;

type NonArrayLike<T> =
  T extends string ? string :
  T extends (infer U)[] ? U :
  T;

contains<T>(actual: ArrayLike<T>, expected: NonArrayLike<T>, name?: string): Test;
notContains<T>(actual: ArrayLike<T>, expected: NonArrayLike<T>, name?: string): Test;

in<T>(actual: NonArrayLike<T>, expected: ArrayLike<T>, name?: string): Test;
notIn<T>(actual: NonArrayLike<T>, expected: ArrayLike<T>, name?: string): Test;
```

```javascript
// suceed
this.contains('something', '1');
this.contains(['something', '1', '2', '3'], '1');
this.contains([1, 2, 3, 4], 5);
this.contains([new Date(), new Date(), new Date()], new Date());

// fail
this.contains('something', new Date());

// succeed
this.in('1', 'dasdsad1');
this.in(5, [1, 2, 3, 4]);
this.in(new Date(), [new Date(), new Date(), new Date()]);
    
// fail
this.contains(5, new Date());
```

Reewr committed:
Allow multi dimensional arrays with contains and in

Reewr committed:
Remove unused type

Thhethssmuz committed:
this is also passed as the first parameter

Thhethssmuz committed:
add generics to eq and ne

Thhethssmuz committed:
use overloading for contains, notContains, in and notIn

Thhethssmuz committed:
add all valid overloads for test

Thhethssmuz committed:
consistency

Reewr committed:
Merge pull request #1 from Thhethssmuz/pr

How about this...

Reewr commented:
Merged the pull request of yours
